File: backend/src/schemas/training.py
"""
Pydantic schemas for Training API endpoints.

These schemas define the structure for request/response validation
and serialization for all SAE training-related API operations.
"""

from datetime import datetime
from typing import Optional, Dict, Any, List, Literal
from enum import Enum

from pydantic import BaseModel, Field, field_validator

from ..models.training import TrainingStatus


class SAEArchitectureType(str, Enum):
    """SAE architecture types."""
    STANDARD = "standard"
    SKIP = "skip"
    TRANSCODER = "transcoder"


class TrainingHyperparameters(BaseModel):
    """Training hyperparameters schema."""

    # SAE Architecture
    hidden_dim: int = Field(..., gt=0, description="Hidden dimension (input/output size)")
    latent_dim: int = Field(..., gt=0, description="Latent dimension (SAE width)")
    architecture_type: SAEArchitectureType = Field(
        SAEArchitectureType.STANDARD,
        description="SAE architecture: standard, skip, or transcoder"
    )

    # Layer configuration
    training_layers: List[int] = Field(
        default=[0],
        min_length=1,
        description="List of layer indices to train SAEs on (e.g., [0, 6, 12])"
    )

    # Sparsity
    l1_alpha: float = Field(
        ...,
        gt=0.00001,
        le=100.0,
        description="L1 sparsity penalty coefficient (SAELens standard: 1.0-10.0 with activation normalization)"
    )
    target_l0: Optional[float] = Field(
        None,
        gt=0,
        le=0.2,
        description="Target L0 sparsity (fraction of active features, typically 0.01-0.05)"
    )
    top_k_sparsity: Optional[float] = Field(
        None,
        gt=0,
        le=1.0,
        description="Top-K sparsity (fraction of neurons to keep active, e.g., 0.05 for 5%). Guarantees exact sparsity level."
    )
    normalize_activations: Optional[str] = Field(
        "constant_norm_rescale",
        description="Activation normalization method: 'constant_norm_rescale' (SAELens standard) or 'none'"
    )

    # Training
    learning_rate: float = Field(..., gt=0, description="Initial learning rate")
    batch_size: int = Field(..., gt=0, description="Training batch size")
    total_steps: int = Field(..., gt=0, description="Total training steps")
    warmup_steps: int = Field(0, ge=0, description="Linear warmup steps")

    # Optimization
    weight_decay: float = Field(0.0, ge=0, description="Weight decay (L2 regularization)")
    grad_clip_norm: Optional[float] = Field(None, gt=0, description="Gradient clipping norm")

    # Checkpointing
    checkpoint_interval: int = Field(1000, gt=0, description="Save checkpoint every N steps")
    log_interval: int = Field(100, gt=0, description="Log metrics every N steps")

    # Dead neuron handling
    dead_neuron_threshold: int = Field(1000, gt=0, description="Steps before a neuron is considered dead")
    resample_dead_neurons: bool = Field(True, description="Resample dead neurons during training")
    resample_interval: int = Field(5000, gt=0, description="Resample dead neurons every N steps")

    @field_validator("training_layers")
    @classmethod
    def validate_training_layers(cls, v: List[int]) -> List[int]:
        """Validate training_layers array."""
        if not v:
            raise ValueError("training_layers must contain at least one layer")
        if any(layer < 0 for layer in v):
            raise ValueError("All layer indices must be non-negative")
        if len(v) != len(set(v)):
            raise ValueError("training_layers must not contain duplicate layer indices")
        return sorted(v)  # Return sorted list for consistency

    model_config = {
        "json_schema_extra": {
            "example": {
                "hidden_dim": 768,
                "latent_dim": 16384,
                "architecture_type": "standard",
                "training_layers": [0, 6, 12],
                "l1_alpha": 0.001,
                "target_l0": 0.05,
                "normalize_activations": "constant_norm_rescale",
                "learning_rate": 0.0003,
                "batch_size": 4096,
                "total_steps": 100000,
                "warmup_steps": 1000,
                "checkpoint_interval": 5000,
                "log_interval": 100
            }
        }
    }


class TrainingCreate(BaseModel):
    """Schema for creating a new training job."""

    model_id: str = Field(..., min_length=1, description="Model ID to train SAE on")
    dataset_id: str = Field(..., min_length=1, description="Dataset ID for training data")
    extraction_id: Optional[str] = Field(None, description="Activation extraction ID (if using pre-extracted activations)")
    hyperparameters: TrainingHyperparameters = Field(..., description="Training hyperparameters")

    @field_validator("model_id")
    @classmethod
    def validate_model_id(cls, v: str) -> str:
        """Validate model ID format."""
        if not v.startswith("m_"):
            raise ValueError("model_id must start with 'm_'")
        return v

    @field_validator("extraction_id")
    @classmethod
    def validate_extraction_id(cls, v: Optional[str]) -> Optional[str]:
        """Validate extraction ID format."""
        if v is not None and not v.startswith("ext_m_"):
            raise ValueError("extraction_id must start with 'ext_m_'")
        return v


class TrainingUpdate(BaseModel):
    """Schema for updating a training job (primarily for status changes)."""

    status: Optional[TrainingStatus] = Field(None, description="Training status")
    progress: Optional[float] = Field(None, ge=0, le=100, description="Training progress")
    current_step: Optional[int] = Field(None, ge=0, description="Current training step")
    current_loss: Optional[float] = Field(None, description="Current reconstruction loss")
    current_l0_sparsity: Optional[float] = Field(None, description="Current L0 sparsity")
    current_dead_neurons: Optional[int] = Field(None, ge=0, description="Current dead neuron count")
    current_learning_rate: Optional[float] = Field(None, ge=0, description="Current learning rate")
    error_message: Optional[str] = Field(None, description="Error message if failed")
    error_traceback: Optional[str] = Field(None, description="Error traceback for debugging")


class TrainingResponse(BaseModel):
    """Schema for training job response."""

    id: str = Field(..., description="Training job ID (format: train_{uuid})")
    model_id: str = Field(..., description="Model ID")
    dataset_id: str = Field(..., description="Dataset ID")
    extraction_id: Optional[str] = Field(None, description="Extraction ID if using pre-extracted activations")

    status: TrainingStatus = Field(..., description="Current training status")
    progress: float = Field(..., description="Training progress (0-100)")
    current_step: int = Field(..., description="Current training step")
    total_steps: int = Field(..., description="Total training steps")

    hyperparameters: Dict[str, Any] = Field(..., description="Training hyperparameters")

    # Current metrics
    current_loss: Optional[float] = Field(None, description="Current reconstruction loss")
    current_l0_sparsity: Optional[float] = Field(None, description="Current L0 sparsity")
    current_dead_neurons: Optional[int] = Field(None, description="Current dead neuron count")
    current_learning_rate: Optional[float] = Field(None, description="Current learning rate")

    # Error handling
    error_message: Optional[str] = Field(None, description="Error message if failed")

    # Paths
    checkpoint_dir: Optional[str] = Field(None, description="Checkpoint directory path")
    logs_path: Optional[str] = Field(None, description="Logs file path")

    # Celery
    celery_task_id: Optional[str] = Field(None, description="Celery task ID")

    # Timestamps
    created_at: datetime = Field(..., description="Job creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    started_at: Optional[datetime] = Field(None, description="Training start timestamp")
    completed_at: Optional[datetime] = Field(None, description="Training completion timestamp")

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "example": {
                "id": "train_abc123",
                "model_id": "m_model123",
                "dataset_id": "ds_dataset456",
                "extraction_id": None,
                "status": "running",
                "progress": 45.5,
                "current_step": 45500,
                "total_steps": 100000,
                "hyperparameters": {
                    "hidden_dim": 768,
                    "latent_dim": 16384,
                    "l1_alpha": 0.001,
                    "learning_rate": 0.0003
                },
                "current_loss": 0.0234,
                "current_l0_sparsity": 0.05,
                "current_dead_neurons": 42,
                "current_learning_rate": 0.00028
            }
        }
    }


class TrainingListResponse(BaseModel):
    """Schema for paginated list of training jobs."""

    data: List[TrainingResponse] = Field(..., description="List of training jobs")
    pagination: Dict[str, Any] = Field(..., description="Pagination metadata")
    status_counts: Dict[str, int] = Field(..., description="Count of trainings by status (all, running, completed, failed)")


class TrainingMetricResponse(BaseModel):
    """Schema for a single training metric record."""

    id: int = Field(..., description="Metric record ID")
    training_id: str = Field(..., description="Training job ID")
    step: int = Field(..., description="Training step")
    timestamp: datetime = Field(..., description="Timestamp")

    # Loss metrics
    loss: float = Field(..., description="Total reconstruction loss")
    loss_reconstructed: Optional[float] = Field(None, description="Reconstruction component")
    loss_zero: Optional[float] = Field(None, description="Zero ablation loss")

    # Sparsity metrics
    l0_sparsity: Optional[float] = Field(None, description="L0 sparsity")
    l1_sparsity: Optional[float] = Field(None, description="L1 sparsity penalty")
    dead_neurons: Optional[int] = Field(None, description="Dead neuron count")

    # Training dynamics
    learning_rate: Optional[float] = Field(None, description="Learning rate")
    grad_norm: Optional[float] = Field(None, description="Gradient norm")

    # Resource metrics
    gpu_memory_used_mb: Optional[float] = Field(None, description="GPU memory usage (MB)")
    samples_per_second: Optional[float] = Field(None, description="Training throughput")

    model_config = {
        "from_attributes": True
    }


class TrainingMetricsListResponse(BaseModel):
    """Schema for list of training metrics."""

    data: List[TrainingMetricResponse] = Field(..., description="List of training metrics")
    pagination: Optional[Dict[str, Any]] = Field(None, description="Pagination metadata")


class CheckpointResponse(BaseModel):
    """Schema for checkpoint response."""

    id: str = Field(..., description="Checkpoint ID (format: ckpt_{uuid})")
    training_id: str = Field(..., description="Training job ID")
    step: int = Field(..., description="Training step at checkpoint")

    loss: float = Field(..., description="Loss at checkpoint")
    l0_sparsity: Optional[float] = Field(None, description="L0 sparsity at checkpoint")

    storage_path: str = Field(..., description="Path to .safetensors file")
    file_size_bytes: Optional[int] = Field(None, description="File size in bytes")

    is_best: bool = Field(..., description="Whether this is the best checkpoint")
    extra_metadata: Optional[Dict[str, Any]] = Field(None, description="Additional metadata")

    created_at: datetime = Field(..., description="Checkpoint creation timestamp")

    model_config = {
        "from_attributes": True
    }


class CheckpointListResponse(BaseModel):
    """Schema for list of checkpoints."""

    data: List[CheckpointResponse] = Field(..., description="List of checkpoints")
    pagination: Optional[Dict[str, Any]] = Field(None, description="Pagination metadata")


class TrainingControlRequest(BaseModel):
    """Schema for training control operations (pause/resume/stop)."""

    action: Literal["pause", "resume", "stop"] = Field(..., description="Control action to perform")


class TrainingControlResponse(BaseModel):
    """Schema for training control response."""

    success: bool = Field(..., description="Whether the action succeeded")
    training_id: str = Field(..., description="Training job ID")
    action: str = Field(..., description="Action that was performed")
    status: TrainingStatus = Field(..., description="New training status")
    message: Optional[str] = Field(None, description="Additional message")

File: frontend/src/types/training.ts
/**
 * Training Types
 *
 * TypeScript interfaces for SAE Training feature.
 * Matches backend Pydantic schemas in src/schemas/training.py and src/models/training.py
 *
 * Backend API Contract:
 * - GET /api/v1/trainings - List all training jobs
 * - GET /api/v1/trainings/:id - Get training job details
 * - POST /api/v1/trainings - Create and start new training job
 * - PATCH /api/v1/trainings/:id - Update training job
 * - DELETE /api/v1/trainings/:id - Delete training job
 * - POST /api/v1/trainings/:id/control - Control training (pause/resume/stop)
 * - GET /api/v1/trainings/:id/metrics - Get training metrics
 * - GET /api/v1/trainings/:id/checkpoints - List checkpoints
 * - GET /api/v1/trainings/:id/checkpoints/best - Get best checkpoint
 *
 * WebSocket Events:
 * - Channel: trainings/{training_id}/progress
 *   - training:created - Job created
 *   - training:progress - Progress update (every 100 steps)
 *   - training:status_changed - Status changed (pause/resume/stop)
 *   - training:completed - Training completed
 *   - training:failed - Training failed
 * - Channel: trainings/{training_id}/checkpoints
 *   - checkpoint:created - Checkpoint saved
 *
 * Status transitions: pending -> initializing -> running -> completed/failed/cancelled
 * Can pause from running -> paused, resume from paused -> running
 */

/**
 * Training job status.
 * Matches backend TrainingStatus enum.
 */
export enum TrainingStatus {
  PENDING = 'pending',
  INITIALIZING = 'initializing',
  RUNNING = 'running',
  PAUSED = 'paused',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

/**
 * SAE training framework types.
 * Each framework has its own optimizer settings, loss formulation,
 * and tailored hyperparameters based on published papers.
 *
 * Matches backend SAEArchitectureType enum.
 */
export enum SAEArchitectureType {
  STANDARD_SAELENS = 'standard_saelens',    // Bricken et al. 2023 — L1 penalty, constant_norm_rescale
  STANDARD_ANTHROPIC = 'standard_anthropic', // Templeton et al. 2024 — L1 penalty, anthropic_rescale, L1~5.0
  JUMPRELU = 'jumprelu',                     // Rajamanoharan et al. 2024 — L0 via STE, learnable thresholds
  TOPK = 'topk',                             // Gao et al. 2024 — structural sparsity, no penalty, aux loss
  SKIP = 'skip',                             // Community variant — L1 + residual skip connection
  TRANSCODER = 'transcoder',                 // Dunefsky et al. 2024 — predicts MLP output from MLP input
  STANDARD = 'standard',                     // Backward compat alias → maps to standard_saelens
}

/**
 * Training hyperparameters configuration.
 * Matches backend TrainingHyperparameters schema.
 */
/**
 * Activation hook types for SAE training.
 * Determines which activation sites to capture during training.
 */
export type HookType = 'residual' | 'mlp' | 'attention';

export interface HyperparametersConfig {
  // SAE Architecture
  /** Hidden dimension (input/output size) */
  hidden_dim: number;
  /** Latent dimension (SAE width, typically 8-32x hidden_dim) */
  latent_dim: number;
  /** SAE architecture type */
  architecture_type: SAEArchitectureType;

  // Layer configuration
  /** List of layer indices to train SAEs on (e.g., [0, 6, 12]) */
  training_layers: number[];
  /** Hook types to train SAEs on (default: ['residual']). Multiple hook types create separate SAEs per layer/hook combination. */
  hook_types?: HookType[];

  // Sparsity (L1-based frameworks: standard_saelens, standard_anthropic, skip, transcoder)
  /** L1 sparsity penalty coefficient. Optional — not used by TopK. */
  l1_alpha?: number;
  /** Target L0 sparsity (fraction of active features, 0-1) */
  target_l0?: number;
  ste_bandwidth?: number;
  /** Mixture over the SELECTED EXTRACTIONS, positional — one weight per entry of
   *  `extraction_ids`, in that order. Normalised server-side. Omit for
   *  proportional-to-real-tokens. Only honoured on the cached-extraction path. */
  dataset_weights?: number[] | null;
  /** Fraction of DOCUMENTS held out of training for out-of-sample evaluation.
   *  0 reproduces the historical behaviour, where every reported number is
   *  in-sample. */
  holdout_fraction?: number;
  /** Held-out tokens evaluated per layer at every log step. Omit for 100,000. */
  holdout_eval_tokens?: number | null;
  /** Held-out tokens per evaluation forward pass. Omit for 2,048. */
  holdout_eval_chunk_tokens?: number | null;
  /** Measure what the SAE reconstruction costs the model after training (default true). */
  evaluate_ce_delta?: boolean;
  /** Real tokens the post-run evaluation reads; 0 skips it. Omit for 131,072. */
  evaluation_token_budget?: number | null;
  /** Random seed for weight init, threshold calibration and token sampling. Omit to derive one from the training id. */
  seed?: number | null;
  /** @deprecated Use top_k (integer count) instead. Percentage-based TopK sparsity. */
  top_k_sparsity?: number;

  // TopK-specific parameters (Gao et al. 2024)
  /** Number of active features per sample for TopK architecture */
  top_k?: number;
  /** Number of dead features for auxiliary loss (default: top_k) */
  aux_k?: number;
  /** Auxiliary dead feature loss coefficient (default: 1/32) */
  aux_loss_alpha?: number;
  /** Adam optimizer epsilon (TopK uses 6.25e-10) */
  adam_epsilon?: number;

  // JumpReLU-specific parameters (Rajamanoharan et al. 2024)
  /** Initial threshold value for JumpReLU activation (default: 0.5) */
  initial_threshold?: number;
  /** KDE bandwidth for STE gradient estimation in JumpReLU (default: 0.01) */
  bandwidth?: number;
  /** L0 sparsity coefficient for JumpReLU (default: 1e-4). */
  sparsity_coeff?: number;
  /** Whether to normalize decoder columns to unit norm */
  normalize_decoder?: boolean;
  /** Activation normalization method (constant_norm_rescale, anthropic_rescale, none) */
  normalize_activations?: string;

  // Training
  /** Initial learning rate */
  learning_rate: number;
  /** Training batch size */
  batch_size: number;
  /** Total training steps */
  total_steps: number;
  /** Linear warmup steps for learning rate */
  warmup_steps?: number;
  /**
   * Linear learning-rate decay to 0 over the final N training steps. 0 (the
   * default) keeps the rate constant after warmup. warmup_steps + lr_decay_steps
   * must not exceed total_steps.
   */
  lr_decay_steps?: number;
  /** Sparsity warmup steps: linearly ramp L1/L0 penalty from 0 to full value. Prevents dead neurons. */
  sparsity_warmup_steps?: number;

  // Optimization
  /** Weight decay (L2 regularization) */
  weight_decay?: number;
  /** Gradient clipping norm */
  grad_clip_norm?: number;

  // Checkpointing
  /** Save checkpoint every N steps */
  checkpoint_interval?: number;
  /** Log metrics every N steps */
  log_interval?: number;

  // Dead neuron handling
  /** Steps before neuron considered dead */
  dead_neuron_threshold?: number;
  /** Resample dead neurons during training */
  resample_dead_neurons?: boolean;
  /** Resample dead neurons every N steps */
  resample_interval?: number;
}

/**
 * Training job.
 * Matches backend Training model and TrainingResponse schema.
 */
export interface Training {
  /** Training job ID (format: train_{uuid}) */
  id: string;
  /** Model ID being trained on */
  model_id: string;
  /** Dataset IDs for training data (supports multiple datasets) */
  dataset_ids: string[];
  /** Primary dataset ID (backward compat, first in dataset_ids) */
  dataset_id: string;
  /** Activation extraction ID (backward compat, first in extraction_ids) */
  extraction_id?: string | null;
  /** Extraction IDs for cached activations (one per dataset) */
  extraction_ids?: string[] | null;
  /** What the run asked for: "auto", "all" or a GPU UUID (null on runs before 2026-09-13 = auto) */
  gpu_request?: string | null;
  /** The GPU (UUID) the run actually ran on, recorded when it started (the first card of a split) */
  gpu_uuid?: string | null;
  /** Every GPU (UUID), in fill order, when the base model was split across GPUs; null otherwise */
  gpu_uuids?: string[] | null;

  // Status and progress
  /** Current training status */
  status: TrainingStatus;
  /** Training progress (0-100) */
  progress: number;
  /** Current training step */
  current_step: number;
  /** Total planned training steps */
  total_steps: number;
  /**
   * Checkpoint step this run was finalized from after being stopped early.
   * null/undefined when the run completed normally. status is 'completed' in
   * both cases, so this is what distinguishes a full run from a salvaged one.
   */
  finalized_from_step?: number | null;

  // Hyperparameters
  /** Training hyperparameters */
  hyperparameters: HyperparametersConfig;

  // Current metrics (latest values)
  /** Current total loss */
  current_loss?: number | null;
  /** Current reconstruction loss component */
  current_reconstruction_loss?: number | null;
  /** Current L1 loss component (l1_alpha * l1_penalty) */
  current_l1_loss?: number | null;
  /** Current L1 alpha (sparsity coefficient, may be warming up) */
  current_l1_alpha?: number | null;
  /** Current L0 sparsity (fraction of active features) */
  current_l0_sparsity?: number | null;
  /** LEGACY FVU: var(x - x_hat) / var(x) with one global mean. Reads low; see utils/fvu. */
  current_fvu?: number | null;
  /** The headline FVU (per-dimension centred). Null for runs recorded before 2026-09-15. */
  current_fvu_centred?: number | null;
  /** Current dead neuron count */
  current_dead_neurons?: number | null;
  /** Current learning rate */
  current_learning_rate?: number | null;

  /** Post-run evaluation on blocks the training never read; null when never evaluated. */
  evaluation?: TrainingEvaluation | null;

  // Error handling
  /** Error message if failed */
  error_message?: string | null;

  // Paths
  /** Checkpoint directory path */
  checkpoint_dir?: string | null;
  /** Logs file path */
  logs_path?: string | null;

  // Celery
  /** Celery task ID */
  celery_task_id?: string | null;

  // Timestamps
  /** Job creation timestamp */
  created_at: string;
  /** Last update timestamp */
  updated_at: string;
  /** Training start timestamp */
  started_at?: string | null;
  /** Training completion timestamp */
  completed_at?: string | null;
}

/**
 * Training metric record (time-series data point).
 * Matches backend TrainingMetric model and TrainingMetricResponse schema.
 */
export interface TrainingMetric {
  /** Metric record ID */
  id: number;
  /** Training job ID */
  training_id: string;
  /** Training step */
  step: number;
  /** Metric collection timestamp */
  timestamp: string;

  // Loss metrics
  /** Total reconstruction loss */
  loss: number;
  /** Reconstruction component of loss */
  loss_reconstructed?: number | null;
  /** Zero ablation loss */
  loss_zero?: number | null;

  // Which rows these are (review R1-C and R1-A A5)
  /** null = the aggregate over every SAE; a layer = that SAE; negative = held-out (-1 - layer) or legacy spliced-CE (-1000 - layer) rows. */
  layer_idx?: number | null;
  /** The SAE's hook type on per-SAE and held-out rows; null on the aggregate and on rows before 2026-09-15. */
  hook_type?: string | null;

  // Sparsity metrics
  /** L0 sparsity (fraction of active features) */
  l0_sparsity?: number | null;
  /** L1 sparsity penalty */
  l1_sparsity?: number | null;
  /** Dead neuron count */
  dead_neurons?: number | null;

  // Reconstruction quality
  /** Legacy global-mean FVU. */
  fvu?: number | null;
  /** Per-dimension-centred FVU (the headline); null on rows written before 2026-09-15. */
  fvu_centred?: number | null;

  // Training dynamics
  /** Learning rate */
  learning_rate?: number | null;
  /** Gradient norm */
  grad_norm?: number | null;

  // Resource metrics
  /** GPU memory usage in MB */
  gpu_memory_used_mb?: number | null;
  /** Training throughput (samples/sec) */
  samples_per_second?: number | null;
}

/**
 * Training checkpoint.
 * Matches backend Checkpoint model and CheckpointResponse schema.
 */
export interface Checkpoint {
  /** Checkpoint ID (format: ckpt_{uuid}) */
  id: string;
  /** Training job ID */
  training_id: string;
  /** Training step at checkpoint */
  step: number;

  // Metrics at checkpoint
  /** Loss at checkpoint */
  loss: number;
  /** L0 sparsity at checkpoint */
  l0_sparsity?: number | null;

  // File storage
  /** Path to .safetensors file */
  storage_path: string;
  /** Checkpoint file size in bytes */
  file_size_bytes?: number | null;

  // Checkpoint metadata
  /** Whether this is the best checkpoint */
  is_best: boolean;
  /** Additional checkpoint metadata */
  extra_metadata?: Record<string, any> | null;

  // Timestamp
  /** Checkpoint creation timestamp */
  created_at: string;
}

/**
 * Training creation request.
 * Matches backend TrainingCreate schema.
 */
export interface TrainingCreateRequest {
  /** Model ID to train SAE on */
  model_id: string;
  /** Dataset IDs for training data (supports multiple datasets) */
  dataset_ids: string[];
  /** Extraction IDs for cached activations (one per dataset) */
  extraction_ids?: string[];
  /** Training hyperparameters */
  hyperparameters: HyperparametersConfig;
  /** `"auto"` (the card with the most free memory when the job starts) or a GPU UUID. */
  gpu?: string;
}

/**
 * Training control request.
 * Matches backend TrainingControlRequest schema.
 */
export interface TrainingControlRequest {
  /**
   * Control action to perform.
   * 'stop' cancels the run; 'stop_and_finalize' also writes community_format
   * from the newest checkpoint so the SAE stays importable.
   */
  action: 'pause' | 'resume' | 'stop' | 'stop_and_finalize';
}

/**
 * Training control response.
 * Matches backend TrainingControlResponse schema.
 */
export interface TrainingControlResponse {
  /** Whether the action succeeded */
  success: boolean;
  /** Training job ID */
  training_id: string;
  /** Action that was performed */
  action: string;
  /** New training status */
  status: TrainingStatus;
  /** Additional message */
  message?: string;
}

/**
 * Paginated training list response.
 * Matches backend TrainingListResponse schema.
 */
export interface TrainingListResponse {
  /** List of training jobs */
  data: Training[];
  /** Pagination metadata */
  pagination: {
    total: number;
    page: number;
    limit: number;
    total_pages: number;
  };
  /** Status counts for filtering */
  status_counts: {
    all: number;
    running: number;
    completed: number;
    failed: number;
  };
}

/**
 * Training metrics list response.
 * Matches backend TrainingMetricsListResponse schema.
 */
export interface TrainingMetricsListResponse {
  /** List of training metrics */
  data: TrainingMetric[];
  /** Pagination metadata (optional) */
  pagination?: {
    total: number;
    page: number;
    limit: number;
    total_pages: number;
  };
}

/**
 * Checkpoint list response.
 * Matches backend CheckpointListResponse schema.
 */
export interface CheckpointListResponse {
  /** List of checkpoints */
  data: Checkpoint[];
  /** Pagination metadata (optional) */
  pagination?: {
    total: number;
    page: number;
    limit: number;
    total_pages: number;
  };
}

/** Status of a training's post-run evaluation. */
export type TrainingEvaluationStatus = 'pending' | 'running' | 'completed' | 'failed' | 'skipped' | 'cancelled';

/** One layer's result: what splicing its SAE costs the model. */
export interface TrainingEvaluationLayer {
  layer: number;
  hook_type?: string;
  tokens?: number;
  ce_spliced: number | null;
  ce_mean_ablated: number | null;
  /** A uniform-output floor on these models: reference only. */
  ce_zero_ablated: number | null;
  ce_delta: number | null;
  /** The headline: (mean-ablated - spliced) / (mean-ablated - base). */
  loss_recovered_vs_mean: number | null;
  loss_recovered_vs_zero: number | null;
  /** KL(untouched model || spliced), nats per predicted token. */
  kl: number | null;
  l0: number | null;
  fvu_centred: number | null;
  fvu_legacy?: number | null;
}

export interface TrainingEvaluationSource {
  label: string;
  dataset_path?: string;
  total_rows?: number;
  rows_read_by_training?: number | null;
  unseen_rows?: number;
  weight?: number;
  blocks_evaluated?: number;
}

/**
 * Where a RUNNING evaluation is (backend `training_evaluation.run_evaluation`): it
 * records `loading_model` before the base model loads, then `means` and
 * `cross_entropy` with batch counts at most once a heartbeat.
 */
export interface TrainingEvaluationProgress {
  stage: 'loading_model' | 'means' | 'cross_entropy' | (string & {});
  batches_done?: number;
  batches?: number;
}

/** trainings.evaluation — see backend services/training_evaluation.py. */
export interface TrainingEvaluation {
  version?: number;
  status: TrainingEvaluationStatus;
  trigger?: 'post_run' | 'rerun';
  reason?: string | null;
  task_id?: string | null;
  /** The running stage; absent before the evaluation starts, and kept on the finished record. */
  progress?: TrainingEvaluationProgress | null;
  /** The hook the SAEs were spliced at: always "resid_post" for a recorded evaluation. */
  hook_point?: string;
  /** What a re-run asked for ("auto", "all" or a GPU UUID), written by POST /trainings/{id}/evaluate. */
  gpu_request?: string | null;
  /** The card a re-run was placed on, as the worker described it (e.g. "GPU 1 (RTX 3090, ...)"). */
  placement?: string;
  requested_at?: string;
  started_at?: string;
  completed_at?: string;
  updated_at?: string;
  config?: { token_budget?: number | null; batch_tokens?: number; seed?: number };
  sources?: TrainingEvaluationSource[];
  skipped_saes?: { layer: number; hook_type: string; reason: string }[];
  tokens?: number;
  predicted_tokens?: number;
  ce_base?: number | null;
  layers?: TrainingEvaluationLayer[];
  all_layers_spliced?: { layers: number[]; ce: number | null; ce_delta: number | null; kl: number | null };
  zero_ablation_note?: string;
}

/**
 * WebSocket training progress event payload.
 */
export interface TrainingProgressEvent {
  training_id: string;
  current_step: number;
  total_steps: number;
  progress: number;
  loss: number;
  reconstruction_loss?: number | null;
  l1_loss?: number | null;
  l1_alpha?: number | null;
  l0_sparsity: number;
  /** Fraction of Variance Unexplained: 0 = perfect reconstruction, 1 = mean-only.
   *  Optional because only architectures that compute it (JumpReLU) report one. */
  fvu?: number | null;
  /** The headline, per-dimension-centred FVU; `fvu` above is the legacy global-mean value. */
  fvu_centred?: number | null;
  dead_neurons: number;
  learning_rate: number;
}

/**
 * WebSocket checkpoint created event payload.
 */
export interface CheckpointCreatedEvent {
  training_id: string;
  checkpoint_id: string;
  step: number;
  loss: number;
  is_best: boolean;
  storage_path: string;
}

/**
 * Result of GET /trainings/{id}/checkpoints/prune-preview.
 * Read-only report of what the retention policy WOULD delete.
 */
export interface CheckpointPrunePreview {
  training_id: string;
  policy: {
    enabled: boolean;
    dry_run: boolean;
    keep_last: number;
    keep_best: boolean;
    min_age_hours: number;
  };
  prunable_steps: number[];
  kept_steps: number[];
  checkpoint_count: number;
  estimated_bytes: number;
  skipped_reason: string | null;
}
